workers/unity/Assets/Gdk/Core/Logging/ForwardingDispatcher.cs

```diff
@@ -0,0 +1,100 @@
+using System.Collections.Generic;
+using Improbable.Worker;
+using UnityEngine;
+
+namespace Improbable.Gdk.Core
+{
+    public class ForwardingDispatcher : ILogDispatcher
+    {
+        private Connection connection;
+
+        private readonly LoggingDispatcher loggingDispatcher = new LoggingDispatcher();
+
+        private readonly LogLevel minimumLogLevel;
+
+        private static readonly Dictionary<LogType, LogLevel> LogTypeMapping = new Dictionary<LogType, LogLevel>
+        {
+            { LogType.Exception, LogLevel.Fatal },
+            { LogType.Error, LogLevel.Error },
+            { LogType.Assert, LogLevel.Error },
+            { LogType.Warning, LogLevel.Warn },
+            { LogType.Log, LogLevel.Info }
+        };
+
+        public ForwardingDispatcher(LogLevel minimumLogLevel = LogLevel.Warn)
+        {
+            this.minimumLogLevel = minimumLogLevel;
+
+            Application.logMessageReceived += LogCallback;
+        }
+
+        private void LogCallback(string message, string stackTrace, LogType type)
+        {
+            if (type == LogType.Exception)
+            {
+                connection.SendLogMessage(LogLevel.Fatal, connection.GetWorkerId(), $"{message}\n{stackTrace}");
+            }
+        }
+
+        public void SetConnection(Connection connection)
+        {
+            this.connection = connection;
+        }
+
+        public void HandleLog(LogType type, LogEvent logEvent)
+        {
+            loggingDispatcher.HandleLog(type, logEvent);
```

Comment:
The structure of this logging setup feels a little chaotic.
In our current setup, we have a top level `ForwardingDispatcher` which takes logging messages from pre-set places and uses a child `LoggingDispatcher` to log messages locally. It also takes messages logged locally from `Application.logMessageReceived` and forwards those to fabric too.

In the UnitySDK, the setup looked like this:
Every UnitySDK error message is logged locally using e.g. `Debug.LogError()`, a single log forwarder forwards all application logs, from the UnitySDK or elsewhere, to fabric as long as a certain min-log level is exceeded. I find the latter approach more elegant and less error prone (see. potential problem with logging the same message to fabric twice)

Reply:
This approach was discussed, but is not feasible in our case. Using this, we would need to add a callback to `logMessageReceived`, whenever a `Debug.Log` is fired, we can automatically forward it to fabric. However, we lose all context and are not able to send e.g. the entity id with it. 

Therefore, we need to define our own logging dispatcher to be able to send additional information to fabric.

Reply:
Can we not log messages in some well defined format and parse the message string to retrieve the context again? Hacky, I know.

Reply:
this doesn't provide the user with a nice API on how to log messages.
I agree, that it was very nice to directly use Debug.LogError, but it doesn't seem to be the right fit for our use case.

Reply:
I don't see the logging something twice , which only happens when running two workers in the same editor, as a problem. Especially as you can get around it by just not having two workers with log dispatchers that capture all editor output. 

However when running two things in the same editor, capturing Untiy logs won't tell you which worker sent it or let you easily configure on a per worker basis what it does.

Parsing strings to regain context is something I want to avoid as it is incredibly error prone, and almost never a good idea.

Reply:
regarding Unity logs: that can be easily mitigated by parsing in the worker id.
 
What I do agree with and forgot to mention earlier: We shouldn't use a LoggingDispatcher inside another dispatcher. This seems confusing and relies on a certain internal implementation of that dispatcher. Can we just use `Debug.Log` directly?

Reply:
@samiwh The logging something twice refers to my comment above. If we log an exception, the exception is also printed locally and then logged to fabric a second time.

The problem I have with the current approach is that errors that we don't explicitly look for happen silently. When looking at the logs of a deployment, it becomes very hard to find bugs that don't happen locally. Users also need to remember to use our custom logging API instead of  `Debug.Log()` which they are used to using. Whenever I want to log something, I find myself always going for `Debug.Log()` first out of pure habit. Having contextual information is nice but often times, I am just interested in knowing whether a bug exists. Except for cases where e.g. entityids are well known, I wonder how much added benefit there is If the log comes with some entityid or request id, basically just some number.

Reply:
We can additionally use `logMessageReceived` to ensure that all warnings and errors gets forwarded to fabric, but other than that we need to provide the functionality of providing more context to fabric when possible/necesasry

I think it's fine to print it in Unity and log it to fabric. This way we ensure that we send the important messages to fabric, but still keep a complete log file on the user's machine in case we need to debug something

Reply:
One issue I can see is that if we want to forward Exception, Error and Warning logs then the ForwardingDispatcher can't log any of those. So it would have to just use a normal Debug.Log in the console when it receives an error not to cause duplication. An alternative to this would be not to use ForwardingDispatcher when in unity editor and then not log when using the ForwardingDispatcher.

Reply:
I had a look at ways to use Debug.Log and it doesn't seem that there is a way to use it with multiple connections. Also after discussing this with Peter (AAA) he mentioned that he didn't like the behavior that messages are sent to console without explicitly specifying it so he is happy with the proposed implementation.